### ISSUE TITLE
The Attempt to Apply Diff Transformer

### ISSUE THREAD

**Issue by Chestnuts52** [microsoft/unilm#1694]
Thank you for your excellent work! 
I have been experimenting with your Diff Transformer in my work on temporal action reconstruction. But I made some modifications, including removing the rotary position encoding (RoPE). This has led to poor model performance, with significant reconstruction errors that do not decrease during training. I have also tried adjusting the number of groups for the key (k) and value (v) vectors, but the errors remain high. Is the rotary position encoding crucial in the Differential Transformer? 
I would appreciate any guidance you can offer. Thank you very much!

**Comment by Chestnuts52** [comment]
> Hi [@Chestnuts52](https://github.com/Chestnuts52) If the baseline Transformer also removes RoPE, Diff and Transformer can be compared against each other fairly. Diff does not rely on RoPE. Would you paste a simple code snippet here, including your implementation of Diff and the specific config (number of layers, number of heads, etc.), so we can find out the problem together.

Thank you very much for your reply! I will share the modified code snippet with you and hope to get your help. Thank you once again!
class MultiheadDiffAttn(nn.Module):
  
    def __init__(self, embed_dim, depth, num_heads, ):
        super().__init__()
        self.embed_dim = embed_dim
        self.num_heads = num_heads
        
        # arg decoder_kv_attention_heads set to half of Transformer's num_kv_heads if use GQA
        # set to same as num_heads if use normal MHA
        self.num_kv_heads = num_heads // 2       # k和v的head个数
        self.n_rep = self.num_heads // self.num_kv_heads        # q的head数量与k/v的head数量的比值，用于扩展k和v张量
        
        self.head_dim = embed_dim // num_heads // 2      
        self.scaling = self.head_dim ** -0.5

        self.q_proj = nn.Linear(embed_dim, embed_dim, bias=False)
        self.k_proj = nn.Linear(embed_dim, embed_dim // self.n_rep, bias=False)
        self.v_proj = nn.Linear(embed_dim, embed_dim // self.n_rep, bias=False)
        self.out_proj = nn.Linear(embed_dim, embed_dim, bias=False)

        self.lambda_init = lambda_init_fn(depth)   
        self.lambda_q1 = nn.Parameter(torch.zeros(self.head_dim, dtype=torch.float32).normal_(mean=0,std=0.1))
        self.lambda_k1 = nn.Parameter(torch.zeros(self.head_dim, dtype=torch.float32).normal_(mean=0,std=0.1))
        self.lambda_q2 = nn.Parameter(torch.zeros(self.head_dim, dtype=torch.float32).normal_(mean=0,std=0.1))
        self.lambda_k2 = nn.Parameter(torch.zeros(self.head_dim, dtype=torch.float32).normal_(mean=0,std=0.1))

        self.subln = RMSNorm(2 * self.head_dim, eps=1e-5, elementwise_affine=True)   
    
    # rel_pos：旋转位置编码（RoPE）的参数
    def forward(self, x,):
        bsz, tgt_len, embed_dim = x.size()
        src_len = tgt_len

        q = self.q_proj(x)
        k = self.k_proj(x)
        v = self.v_proj(x)

        q = q.view(bsz, tgt_len, 2 * self.num_heads, self.head_dim)
        k = k.view(bsz, src_len, 2 * self.num_kv_heads, self.head_dim)
        v = v.view(bsz, src_len, self.num_kv_heads, 2 * self.head_dim)

        # 旋转位置编码，增强模型对位置信息的感知能力  
        #q = apply_rotary_emb(q, *rel_pos, interleaved=True)
        #k = apply_rotary_emb(k, *rel_pos, interleaved=True)

        q = q.transpose(1, 2)
        k = repeat_kv(k.transpose(1, 2), self.n_rep)
        v = repeat_kv(v.transpose(1, 2), self.n_rep)

        q *= self.scaling     
        attn_weights = torch.matmul(q, k.transpose(-1, -2))  
        attn_weights = F.softmax(attn_weights, dim=-1, dtype=torch.float32).type_as(
            attn_weights
        )

        lambda_1 = torch.exp(torch.sum(self.lambda_q1 * self.lambda_k1, dim=-1).float()).type_as(q)   # A1
        lambda_2 = torch.exp(torch.sum(self.lambda_q2 * self.lambda_k2, dim=-1).float()).type_as(q)   # A2
        lambda_full = lambda_1 - lambda_2 + self.lambda_init   # 差分
        attn_weights = attn_weights.view(bsz, self.num_heads, 2, tgt_len, src_len)  
        attn_weights = attn_weights[:, :, 0] - lambda_full * attn_weights[:, :, 1] 

        attn = torch.matmul(attn_weights, v)     
        attn = self.subln(attn)   
        attn = attn * (1 - self.lambda_init)    
        attn = attn.transpose(1, 2).reshape(bsz, tgt_len, self.num_heads * 2 * self.head_dim)

        attn = self.out_proj(attn)  
        return attn
To maintain consistency with my original multi-head self-attention mechanism, I have set the number of heads to 8 and the number of layers to 8. The head dimensions for k and v are set to half of the number of heads, which is 4. Additionally, since my task requires capturing both preceding and following information, I have removed the causal mask.
I've also experimented with an alternative grouped attention implementation, modifying the expansion operations for k  and v， but the results still haven't changed. I'm hoping to get your valuable insights on this.

**Comment by YTianZHU** [comment]
Hi @Chestnuts52 , maybe my comment in the code causes some misunderstanding. self.num_kv_heads doesn't have to be half of self.num_heads
If your baseline Transformer has 8 heads and 8 kv_heads, then you should give DiffTransformer 4 heads and 4 kv_heads. Because each head in Diff has doubled dimension comparing to one head in baseline Transformer.

**Comment by YTianZHU** [comment]
@Chestnuts52 we add an example file, you can check that to find matching configs
https://github.com/microsoft/unilm/blob/master/Diff-Transformer/example.py

**Comment by Chestnuts52** [comment]
@YTianZHU Thank you very much for your suggestions and guidance! Following your advice on adjusting the number of heads for q,k,v, the training error in my task has become normal. I'm really looking forward to seeing the performance improvements that the Diff Transformer can bring. Thank you once again!
